modules/overrides/user_config_overrides.go

```diff
@@ -0,0 +1,384 @@
+		// Create overrides directory with necessary permissions
+		err = os.MkdirAll(path.Join(cfg.Local.Path, overridesKeyPath), os.ModePerm)
+	case "gcs":
+		reader, writer, _, err = gcs.New(cfg.GCS)
+	case "s3":
+		reader, writer, _, err = s3.New(cfg.S3)
+	case "azure":
+		reader, writer, _, err = azure.New(cfg.Azure)
+	default:
+		err = fmt.Errorf("unknown backend %s", cfg.Backend)
+	}
+	// TODO wrap reader and writer in cache
+	return
+}
+
+func (o *userConfigOverridesManager) starting(ctx context.Context) error {
+	if err := services.StartManagerAndAwaitHealthy(ctx, o.subservices); err != nil {
+		return errors.Wrap(err, "unable to start overrides subservices")
+	}
+
+	return o.reloadAllTenantLimits(ctx)
+}
+
+func (o *userConfigOverridesManager) loop(ctx context.Context) error {
+	ticker := time.NewTicker(o.cfg.PollInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-ticker.C:
+			err := o.reloadAllTenantLimits(ctx)
+			if err != nil {
+				level.Error(o.logger).Log("msg", "failed to refresh user-configurable config", "err", err)
+			}
+			continue
+
+		case err := <-o.subservicesWatcher.Chan():
+			return errors.Wrap(err, "overrides subservice failed")
+		}
+	}
+}
+
+func (o *userConfigOverridesManager) stopping(error) error {
+	return services.StopManagerAndAwaitStopped(context.Background(), o.subservices)
+}
+
+func (o *userConfigOverridesManager) reloadAllTenantLimits(ctx context.Context) error {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "overrides.reloadAllTenantLimits")
+	defer span.Finish()
+
+	traceID, _ := tracing.ExtractTraceID(ctx)
+	level.Info(o.logger).Log("msg", "reloading all tenant limits", "traceID", traceID)
+
+	// TODO can we make this lock smaller? this will block all operations that read from the overrides
+	o.mtx.Lock()
+	defer o.mtx.Unlock()
+
+	// List tenants with user-configurable overrides
+	tenants, err := o.listTenantsUnderLock(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Clean up cached tenants that have been removed from the backend
+	for cachedTenant := range o.tenantLimits {
+		if doesNotContain(tenants, cachedTenant) {
+			delete(o.tenantLimits, cachedTenant)
+		}
+	}
+
+	// For every tenant with user-configurable overrides, download and cache them
+	for _, tenant := range tenants {
+		_, err = o.getTenantLimitsUnderLock(ctx, tenant)
+		if err != nil {
+			// TODO should we keep trying the other tenants and return a combined error message?
+			//  this implementation gives up after a single failure
+			return errors.Wrap(err, "failed to load tenant limits")
+		}
+	}
+
+	return nil
+}
+
+func (o *userConfigOverridesManager) listTenantsUnderLock(ctx context.Context) ([]string, error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "overrides.listTenantsUnderLock")
+	defer span.Finish()
+
+	// TODO to avoid polling the entire bucket, use a tenant list and keep it in a shared cache
+	tenants, err := o.r.List(ctx, []string{overridesKeyPath})
```

Comment:
this is threadsafe and doe not require locking

Reply:
I was thinking of a scenario in which someone sends a POST or DELETE request right after we called list. In this case the list we just fetched would be outdated. I think I should take another look at this and simplify it, we really don't need a constant lock for the whole thing...

Reply:
These kinds of state based race conditions are unavoidable. We should just push them to object storage as much as possible.